xds/src/main/java/io/grpc/xds/sds/SslContextSecretVolumeSecretProvider.java

```diff
@@ -0,0 +1,192 @@
+      return null;
+    }
+    checkArgument(
+        certContext.getTrustedCa().getSpecifierCase() == SpecifierCase.FILENAME,
+        "filename expected");
+    return certContext;
+  }
+
+  @VisibleForTesting
+  @Nullable
+  static TlsCertificate validateTlsCertificate(
+      @Nullable TlsCertificate tlsCertificate, boolean optional) {
+    if (tlsCertificate == null) {
+      checkArgument(optional, "tlsCertificate is required");
+      return null;
+    }
+    if (optional
+        && (tlsCertificate.getPrivateKey().getSpecifierCase() == SpecifierCase.SPECIFIER_NOT_SET)
+        && (tlsCertificate.getCertificateChain().getSpecifierCase()
+            == SpecifierCase.SPECIFIER_NOT_SET)) {
+      return null;
+    }
+    checkArgument(
+        tlsCertificate.getPrivateKey().getSpecifierCase() == SpecifierCase.FILENAME,
+        "filename expected");
+    checkArgument(
+        tlsCertificate.getCertificateChain().getSpecifierCase() == SpecifierCase.FILENAME,
+        "filename expected");
+    return tlsCertificate;
+  }
+
+  static SslContextSecretVolumeSecretProvider getProviderForServer(
+      TlsCertificate tlsCertificate, @Nullable CertificateValidationContext certContext) {
+    // first validate
+    validateTlsCertificate(tlsCertificate, /* optional= */ false);
+    // certContext exists in case of mTLS, else null for a server
+    if (certContext != null) {
+      certContext = validateCertificateContext(certContext, /* optional= */ true);
+    }
+    String privateKeyPassword =
+        tlsCertificate.hasPassword() ? tlsCertificate.getPassword().getInlineString() : null;
+    String trustedCa = certContext != null ? certContext.getTrustedCa().getFilename() : null;
+    return new SslContextSecretVolumeSecretProvider(
+        tlsCertificate.getPrivateKey().getFilename(),
+        privateKeyPassword,
+        tlsCertificate.getCertificateChain().getFilename(),
+        trustedCa,
+        /* server= */ true);
+  }
+
+  static SslContextSecretVolumeSecretProvider getProviderForClient(
+      @Nullable TlsCertificate tlsCertificate, CertificateValidationContext certContext) {
+    // first validate
+    validateCertificateContext(certContext, /* optional= */ false);
+    // tlsCertificate exists in case of mTLS, else null for a client
+    if (tlsCertificate != null) {
+      tlsCertificate = validateTlsCertificate(tlsCertificate, /* optional= */ true);
+    }
+    String privateKey = null;
+    String privateKeyPassword = null;
+    String certificateChain = null;
+    if (tlsCertificate != null) {
+      privateKey = tlsCertificate.getPrivateKey().getFilename();
+      if (tlsCertificate.hasPassword()) {
+        privateKeyPassword = tlsCertificate.getPassword().getInlineString();
+      }
+      certificateChain = tlsCertificate.getCertificateChain().getFilename();
+    }
+    return new SslContextSecretVolumeSecretProvider(
+        privateKey,
+        privateKeyPassword,
+        certificateChain,
```

Comment:
if those can be nullable, please mark the constructor and fields unless this is a bug

Reply:
done